### ISSUE TITLE
Clarifying optional versus required parameters in the AKE section

### ISSUE THREAD

**Issue by kevinlewi** [cfrg/draft-irtf-cfrg-opaque#74]
In this [section on AKE](https://github.com/cfrg/draft-irtf-cfrg-opaque/blame/master/draft-irtf-cfrg-opaque.md#L1008-L1014), we should clear up exactly which parameters are required and which are optional. Then we can also work towards defining a wire format for these AKEs if necessary.

A couple of concrete questions:
1) Can we simply set idU = static pkU and idS = static pkS?
2) Can we remove mention of all of the following: info1*, info2*, Einfo2*, info3*, Einfo3*?

**Comment by chris-wood** [comment]
> 1. Can we simply set idU = static pkU and idS = static pkS?

This had been discussed in the past, and I think it depends on whether or not there exists a secure binding between the public key(s) and corresponding identity. (And that seems mostly external to the protocol.) I need to think about this more. @hugokraw probably also has opinions. :-)

> 2. Can we remove mention of all of the following: info1*, info2*, Einfo2*, info3*, Einfo3*?

These fields exist to make the AKE instantiation more generic, e.g., maybe `info1` contains the client's set of application protocols run on top of the AKE. I think we should remove them, since applications that need this flexibility should be using TLS. I view the 3DH and SIGMA-I variants as minimalist OPAQUE instantiations.

**Comment by hugokraw** [comment]
Regarding point 1: In general, identities refer to logical entities on the basis of which policy decisions by the receiving party are made (do I want to talk to this identity, under what security level, which access rights, etc.) A public key may be used as an identity but most of the time it is not (e.g. , it is the subject in a certificate rather than the public key). In the case of OPAQUE,  using public keys may be more acceptable since those are fixed after registration and the client authenticates the server's public key via the password. I am not sure if this means that the client will never need to make decisions based on the server's logical identity but I do not see an obvious setting where this is the case. However, note that in the three protocols in the draft, the client does not use the user's public key. So if we are going to use it then we either send it as part of the authenticated information in the envelope or require the client to re-compute it from the user's private key at the cost of an exponentiation.

**Comment by hugokraw** [comment]
Regarding point 2: We want these AKEs to be minimalistic but not useless. I do not see the case of any key exchange protocol that does not require sending some additional setting-specific information in its flows. Instead of having these fields named specifically, you may define some extensibility mechanism to transport additional information, but some mechanism for that must exist.

**Comment by kevinlewi** [comment]
Thank you for the comments!

For point 1: I would be in favor of requiring the client to re-compute the user's public key from its private key at the cost of an exponentiation, if it means that we do not need to send idU over the wire for the KE1 message. Similarly, for idS, since the server's public key is already required to be a part of the authenticated envelope, I would be in favor of omitting idS from the KE2 message as well (it is not currently there but just in case that comes up).

For point 2: Fair enough, we can keep them as optional parameters, but just to double-check, for the sake of minimalism: do you think we need both the plaintext variants (info1*, info2*, info3*) in addition to the encrypted variants (Einfo2*, Einfo3*), or would one set of these suffice?

**Comment by chris-wood** [comment]
Regarding point 2, why can't applications send info *after* the AKE completes? (I understand that it burns a round trip, but that seems preferable to additional complexity in the AKE protocol itself.)

**Comment by hugokraw** [comment]
The plaintext info field may be needed to transport information needed to select the algorithms and type of protection applied by the AKE and cannot be postponed to after the end of the AKE. The encrypted info field may be needed to transport information that requires privacy. I am all for simplifying wherever possible but I also want a usable specification. Eliminating all information beyond the core crypt operations does not seem to lead to such usable specification.  

I am ok with letting pkU and pkS act as identities but I can also see implementations that would prefer to use something like a uid@site.dom as idU and save the computation of pkU. Would you be ok with having the id's be optional elements in the envelope (but omitted from the AKE messages)?

**Comment by chris-wood** [comment]
> The plaintext info field may be needed to transport information needed to select the algorithms and type of protection applied by the AKE and cannot be postponed to after the end of the AKE. 

I'm not sure I agree with this. If cryptographic algorithm negotiation is needed for the AKE, then we should specify the mechanics to do that. We shouldn't just provide a generic slot for application info and punt the problem downstream.

> The encrypted info field may be needed to transport information that requires privacy. 

If privacy is required, this information can be sent after the AKE completes. Why does it need to be done in-band?

> I am all for simplifying wherever possible but I also want a usable specification. Eliminating all information beyond the core crypt operations does not seem to lead to such usable specification.

I think it's the opposite. If we keep the specification as minimal as I'm advocating for, then applications are free to use the AKE's output however they want.

**Comment by kevinlewi** [comment]
> Would you be ok with having the id's be optional elements in the envelope (but omitted from the AKE messages)?

I believe idU and idS can already be specified as optional elements in the envelope (see [here](https://github.com/cfrg/draft-irtf-cfrg-opaque/blame/master/draft-irtf-cfrg-opaque.md#L438-L439)), and so I would be happy to go with omitting them from the AKE messages.

**Comment by hugokraw** [comment]
> > Would you be ok with having the id's be optional elements in the envelope (but omitted from the AKE messages)?
> 
> I believe idU and idS can already be specified as optional elements in the envelope (see [here](https://github.com/cfrg/draft-irtf-cfrg-opaque/blame/master/draft-irtf-cfrg-opaque.md#L438-L439)), and so I would be happy to go with omitting them from the AKE messages.

Ok. So we can eliminate them from the AKE messages but have an explicit specification that the values idU and idS in the KDF are those specified in the envelope or, if omitted there, then pkU and pkS  (with a note that pkU will need to be computed by the client if not transmitted in the envelope).

**Comment by hugokraw** [comment]
> I think it's the opposite. If we keep the specification as minimal as I'm advocating for, then applications are free to use the AKE's output however they want.

The are always free to use the AKE output however they want. The question is whether they will have enough information to run and complete the AKE if they cannot send more than the bare cryptographic elements during the AKE.

**Comment by chris-wood** [comment]
> The are always free to use the AKE output however they want. The question is whether they will have enough information to run and complete the AKE if they cannot send more than the bare cryptographic elements during the AKE.

As specified currently, and ignoring the `info` fields, is this the case? If so, shouldn't we specify everything that's needed to complete the AKE in this spec? I don't see how we can punt on that!

**Comment by hugokraw** [comment]
> As specified currently, and ignoring the `info` fields, is this the case? If so, shouldn't we specify everything that's needed to complete the AKE in this spec? I don't see how we can punt on that!

I am thinking of these AKEs as something people will need to customize to their application/setting. The usage by Facebook and the ones by AWS are very different and they will need to fill in the details according to their needs.

**Comment by hugokraw** [comment]
If it helps (I am not fully convinced), I am ok with removing info and einfo in the current description, and instead adding language saying that what is shown is the minimal cryptographic skeleton and that applications may need to transport additional information as part of the AKE. In such a case, any information transmitted as part of the AKE messages must be included in the transcript fields (that are mac-ed in 3DH/HMQV and signed in SIGMA).  In addition, we specify the derivation of authenticated encryption keys (denote Ke2, Ke3) to be used for confidentiality protection of elements in the transcript that require such protection.

**Comment by kevinlewi** [comment]
Since the discussion on point #2 seems to be around either: leaving in the optional parameters info and einfo, or omitting them entirely, it seems like the tradeoffs are the following:

Pros for leaving in the optional parameters info* and einfo*:
- A potential application that wants to use these fields can do so without having to wait the extra round trip for AKE to finish

Cons for leaving in the optional parameters info* and einfo*:
- Increased spec / API complexity for implementers
- Increased bandwidth due to bytes spent on length-prefixing the optional parameters in AKE even if they are not used

Is this an accurate summary? If so, I am leaning slightly towards keeping in the optional parameters, but not with a strong preference :)

**Comment by chris-wood** [comment]
> Is this an accurate summary? If so, I am leaning slightly towards keeping in the optional parameters, but not with a strong preference :)

Yeah, I think that's mostly accurate. Can you elaborate on why you're leaning that way? (If you used them in your implementation, what were the values of the fields?)

**Comment by hugokraw** [comment]
My feeling is that we serve better the potential implementers of OPAQUE that need to embed the protocol in their specific application by guiding them toward they way of adding application specific elements to the protocol, in particular telling them that they should add these elements to the authenticated transcript and use Ke to protect confidentiality if needed. As I wrote elsewhere, I am ok if we remove info, einfo from the spec but then we must have English text about the need to include such elements under the authenticated transcript. Also noting the potential need for Ke (which we would include in the key derivation even if we do not show einfo* explicitly).  My bottom line: We would be serving implementers of OPAQUE in specific applications better with the info fields than without. (But who am I to tell what implementers really need? :-)

**Comment by chris-wood** [comment]
> My bottom line: We would be serving implementers of OPAQUE in specific applications better with the info fields than without.

Yeah, I think if the goal is to allow applications to build a one-off AKE specific to their use case, then I think this is the right call. If, on the other hand, we're specifying a feature-complete AKE in this spec, I lean the other way.

**Comment by kevinlewi** [comment]
> > Is this an accurate summary? If so, I am leaning slightly towards keeping in the optional parameters, but not with a strong preference :)
> 
> Yeah, I think that's mostly accurate. Can you elaborate on why you're leaning that way? (If you used them in your implementation, what were the values of the fields?)

One example that comes to mind is the following: Suppose a client is using a password to log in to a server, and on the KE2 message, the server wants to send to the client auxiliary information about the last time the client successfully logged in (IP address, # of incorrect login attempts since last login, etc.). These are of course things that can wait 1 more round trip for, but in certain applications it may be more convenient to not have to wait.

Anyway, the main reason why I am leaning towards including the optional parameters is because I feel like we are more likely to regret excluding them than the other way around. I believe the added API complexity only affects those who have to implement a spec-compliant OPAQUE (hopefully a small population) and the benefits of having the flexiblity to use these optional parameters would affect those who use the spec-compliant OPAQUE implementations (hopefully a larger population).

@huitseeker in case he has any additional thoughts on this

**Comment by hugokraw** [comment]
> I feel like we are more likely to regret excluding them than the other way around

I would have not dared to show my feelings like Kevin's but I strongly feel like him :-)

**Comment by hugokraw** [comment]
> Yeah, I think if the goal is to allow applications to build a one-off AKE specific to their use case, then I think this is the right call. If, on the other hand, we're specifying a feature-complete AKE in this spec, I lean the other way.

This is indeed the point in which our view diverge. I hear of several people with various interests, including Kevin/Facebook, Crockett/AWS and people from some banks. They were all talking of specific applications they have rather than using a general Web/TLS-based protocol. I think that the document we are writing needs to serve these people. It can also serve those integrating OPAQUE in other protocols IKEv2, SSH, ...

**Comment by huitseeker** [comment]
Optional information parameters can be useful, and if the past is prologue, they'll eventually be useful to shave off round-trips on auxiliary data in what's after all a latency-sensitive initial setup.

So, in that regard, if we keep those useful parameters in:
- might it make sense to be very clear on what guarantees (as to authentication, compromise resistance, resistance to key compromise ...) the server can expect at the time it sends these optional parameters?
- would it make sense to make an opt-in to those parameters a strong signal, maybe by committing to it in a ciphersuite string?

**Comment by chris-wood** [comment]
Yeah, on reflection, I think keeping the parameters is probably the best path forward. @kevinlewi, how do you want to deal with point (1)?

**Comment by kevinlewi** [comment]
For #1, recapping the summary from @hugokraw :

> Ok. So we can eliminate them from the AKE messages but have an explicit specification that the values idU and idS in the KDF are those specified in the envelope or, if omitted there, then pkU and pkS (with a note that pkU will need to be computed by the client if not transmitted in the envelope).

So, let's remove their presence from the AKE messages, and I would vote for setting idU = pkU and idS = pkS when computing the AKE components (instead of reusing idU and idS that were optionally specified in the envelope).

**Comment by chris-wood** [comment]
That seems reasonable to me. Can you please draft a PR?

**Comment by hugokraw** [comment]
> So, let's remove their presence from the AKE messages, and I would vote for setting idU = pkU and idS = pkS when computing the AKE components (instead of reusing idU and idS that were optionally specified in the envelope).

Do you mean removing idU and idS from the envelope, even as optional, and always using pkU and pkS as the identities in the AKE? I am responding to that interpretation. If that's not what you were saying, ignore this.

 Identities are central to authorization, access control, etc. and password protocols are central to these services so I think we need to leave flexibility there. Having the optional idU, idS in the envelope and pkU, pkS as default seems like a good balance between flexibility and ease of implementation.

**Comment by kevinlewi** [comment]
@hugokraw That is not quite what I mean; I am all for keeping idU and idS as optional parameters in the envelope. However, the **AKE portion** also refers to idU and idS: in particular, the values referenced [here](https://github.com/cfrg/draft-irtf-cfrg-opaque/blame/master/draft-irtf-cfrg-opaque.md#L1138-L1139). The question is -- should these simply be set to idU = pkU and idS = pkS, or should they be the same idU and idS that were optionally specified in the envelope?

I am advocating for idU = pkU and idS = pkS just for the **AKE portion**, the values [here](https://github.com/cfrg/draft-irtf-cfrg-opaque/blame/master/draft-irtf-cfrg-opaque.md#L1138-L1139). It would be great to get your thoughts on this!

**Comment by hugokraw** [comment]
There is no use for idU and idS transported in the envelope other than to serve as identities in the AKE in lieu of any other identity that the parties may use to initially identify each other (for the client to know which server to contact and for the client to let the server know what user record to fetch). For example, I use hugokraw as my moniker in github but github may actually use my email address as my official identity (I don't know if this is the case, it is an imaginary example as far as I can tell).  I find it too limiting to dictate applications what's the identity they want to bind the AKE to.

**Comment by chris-wood** [comment]
Summary of discussion from the call: There are two "types" of identities -- one used to establish a connection, and one registered and stored in the envelope. If there is no identity in the envelope, the AKE ID matches that used for the outer connection. Conversely, if there is an identity in the envelope, that is the used in AKE.

@hugokraw, @kevinlewi: does this match your understanding? 

If so, we should elaborate on [this section](https://cfrg.github.io/draft-irtf-cfrg-opaque/draft-irtf-cfrg-opaque.html#name-ake-execution-and-party-ide) to explain the differences between these identities and their respective use cases. In the process, we should recommend one implementation strategy for applications, e.g., by recommending that the AKE ID be the public key, and if a separate identity is desired, it's put in the envelope. In doing so, we should add text to deal with the case where application identities are long -- perhaps too long to store in the envelope. In such cases, a hash of the identity might be stored in the envelope and threaded into the AKE, and the server needs to check the hash against the long-form identity.

**Comment by kevinlewi** [comment]
Excellent -- yes, thank you, this matches my understanding!

Just to double-check, the same logic applies for idS, of course, right? (Default to pkS, unless idS is specified in the envelope, in which case use that)

**Comment by hugokraw** [comment]
Good (and yes, same logic for idS).

Note: Currently, [here](https://github.com/cfrg/draft-irtf-cfrg-opaque/blob/master/draft-irtf-cfrg-opaque.md#finalizerequest) we say "Applications may optionally include pkU, idU, or idS...." in the envelope. We may want to clarify that if pkU is to be used as idU, this value needs to be sent in the envelope or the client would need to compute pkU from skU.

**Comment by kevinlewi** [comment]
@chris-wood @hugokraw Thanks! There is one more edge case to discuss:

Suppose idU and idS are part of the envelope, but they are stored as a secret_credential (encrypted under `rwdU`) as opposed to something that the server can read in plaintext? The server will not be able to perform the AKE key derivation since it cannot read these values in place of idU and idS.

Also, what if they are stored authenticated in the envelope under cleartext_credentials? Do we expect the server to read the values without trying to check the MAC on them (since that would also require `rwdU`)?

**Comment by hugokraw** [comment]
If idU and/or idS are stored under secret_credential, the server needs some other form of access to these values. The server cannot decrypt or authenticate the envelope; only the client in possession of the password (and after computing rwdU) can. Typically, the server will store these values separately or can derive them from other information it has, e.g., it knows its own identity idS or it derives idU from the identity it uses to identify the user's record. If the values are stored in the envelope in cleartext, the server can choose to read them from there; it is an implementation issue.

**Comment by kevinlewi** [comment]
Makes sense. In that case, I propose that we set the default behavior to use idU = pkU and idS = pkS. And then, if the wrapper protocol wishes to use a custom idU and idS, then the OPAQUE protocol can accept these custom parameters from the wrapper protocol and use them in the AKE instead. Thus, the envelope would be completely ignored when doing this calculation.

**Comment by chris-wood** [comment]
If the envelope's identities are (effectively) ignored, would we keep idU and idS as valid credential types?

**Comment by kevinlewi** [comment]
@chris-wood That's a fair question -- perhaps we can create a separate issue to discuss the inclusion of idU and idS in the envelope, once we land the changes prescribed here?

**Comment by hugokraw** [comment]
The question is whether the client needs to accept the identities provided by the wrapper as the client cannot authenticate those. If authentication is needed then one must either use the default pkS identity for S or otherwise idS needs to be authenticated by the envelope. To illustrate the need for authentication, consider a setting where the application running on the client side has an access control policy with some permission rules for accessing a resource R.  Both idS and idS* are known server names to which the user may connect so both names have defined permissions for resource R. Note that the policy is defined in terms of names, not public keys (as with subject names in certificates vs.  public keys), and the names are received from the AKE layer once authentication succeeds (the AkE may pass several elements to the higher layers, including the peer's identity).

Now suppose the client connects to idS who runs a rogue wrapper that provides the client with the correct envelope for idS but gives idS* as the identity.  The client will end the AKE successfully but with identity idS*. Indeed, the server (whose real name is idS) will authenticate using his own private key (corresponding to the public key in the envelope) but will enter the name idS* to the AKE which the client will do too (as it is using the identity provided by the wrapper). So the AKE level will pass to the access control policy the identity idS* and idS (not idS*) will be granted access to resource R.

**Comment by kevinlewi** [comment]
@hugokraw I am not entirely sure if you are proposing that we deviate from my [above comment](https://github.com/cfrg/draft-irtf-cfrg-opaque/issues/74#issuecomment-724214162) (it seems like you are still in agreement), but I think to address the concern that you mentioned, the OPAQUE implementation should simply surface both idS (found in the envelope) and idS* (as output by the AKE) to the client, and allow the wrapper protocol on the client side decide whether or not it is valid for idS != idS*. Similarly for idU / idU* on the server-side.

**Comment by hugokraw** [comment]
If I understand correctly your proposal, it says: "Either use the default identities pkU, pkS, or let the wrapper decide on these identities." This presumes the wrapper application to be trusted and a final decision maker on what identities are acceptable.
In this case, I do not see the need to ever transport identities in the envelope. Or do you still think there is a point in allowing identities in the envelope in this case? How would they be used? 

I don't have an opinion on the trust one should put on the wrapper (I don't have enough use cases in mind to reason about it), but I'd hesitate to specify OPAQUE with such a universal trust assumption. It feels safer  (but maybe too limiting?) to make all decisions based on what OPAQUE can validate. Namely, follow the strict policy by which AKE identities are defined in the envelope and if they are missing from it then they default to pkU and pkS.

**Comment by kevinlewi** [comment]
OK, I think this makes sense, and we can first check for the existence of the identities in the envelope, and if they are missing, default to pkU and pkS (as we agreed to in [this comment](https://github.com/cfrg/draft-irtf-cfrg-opaque/issues/74#issuecomment-723337010)).

But just as an FYI, since we must distinguish between the OPAQUE protocol run on the server versus the wrapper protocol that calls OPAQUE on the server, this means that the OPAQUE server protocol will have to **ask the wrapper protocol for the idU and idS it should use for the AKE**, since it cannot decrypt/authenticate the contents of the envelope. Just want to make sure we are clear on that.

If so, I believe we are good to go.

**Comment by hugokraw** [comment]
There are many ways in which the server can decide on the values of idS and idU. It can indeed ask the wrapper for those values but it could also  set idS and idU at registration (or later) and store them in the user's record, or it can read these values from the envelope if stored unencrypted there.  All these options are legit. 
Note that the setting here is very different for the user and server.  The user has no way of authenticating anything other than with his password (hence it can only authenticate information in the envelope) but the server has local storage with information (e.g., idS and idU) that it can trust.

**Comment by chris-wood** [comment]
@kevinlewi I think we can close this as resolved now. Do you want to do the honors?

**Comment by kevinlewi** [comment]
Thanks everyone for engaging in the discussions!
